src/frontend/src/optimizer/plan_node/logical_hop_window.rs

```diff
@@ -167,42 +165,45 @@ impl ColPrunable for LogicalHopWindow {
             self.input().schema().len(),
         );
         let (new_hop, _) = self.rewrite_with_input(input, input_change);
-        let required_output_cols = {
+        let output_cols = {
             // output cols = { cols required by upstream from input node } ∪ { additional window
             // cols }
-
-            // `required_output_cols` indicates whether a column is needed or not
-            // get indices of columns which are required by upstream in `new_hop`
-            let mut required_output_cols = {
-                // map the indices from output to input
-                let input_required_cols = required_cols
-                    .iter()
-                    .filter_map(|&idx| o2i.try_map(idx))
-                    .collect_vec();
-                // this mapping will only keeps required input columns
-                let mapping = ColIndexMapping::with_remaining_columns(
-                    &input_required_cols,
-                    self.input().schema().len(),
-                );
-                input_required_cols
-                    .iter()
-                    .filter_map(|&idx| mapping.try_map(idx))
-                    .collect_vec()
-            };
-
             let win_start_index = new_hop.schema().len() - 2;
             let win_end_index = new_hop.schema().len() - 1;
-            if require_win_start {
-                required_output_cols.push(win_start_index);
-            }
-            if require_win_end {
-                required_output_cols.push(win_end_index);
+            #[derive(Copy, Clone, Debug)]
+            enum IndexType {
+                Input(usize),
+                WindowStart,
+                WindowEnd,
             }
-            required_output_cols
+            // map the indices from output to input
+            let input_required_cols = required_cols
+                .iter()
+                .map(|&idx| {
+                    if let Some(idx) = o2i.try_map(idx) {
+                        IndexType::Input(idx)
+                    } else if idx == self.window_start_col_idx() {
+                        IndexType::WindowStart
+                    } else {
+                        IndexType::WindowEnd
+                    }
+                })
+                .collect_vec();
```

Comment:
I add an enum `IndexType` here to distinguish between columns that are from upstream operators and columns that are generated by `LogicalHopWindow` itself(WindowStart, WindowEnd). Not sure if there is a better way to implement this.